File: jsimple-oauth/c#/jsimple/oauth/model/ParameterList.cs
using System.Diagnostics;
using System.Collections.Generic;
using System.Text;

namespace jsimple.oauth.model
{

	using OAuthEncoder = jsimple.oauth.utils.OAuthEncoder;
	using PlatformUtils = jsimple.util.PlatformUtils;


	/// <summary>
	/// @author: Pablo Fernandez
	/// </summary>
	public class ParameterList
	{
		private const char QUERY_STRING_SEPARATOR_CHAR = '?';
		private const string QUERY_STRING_SEPARATOR = "?";
		private const string PARAM_SEPARATOR = "&";
		private const string PAIR_SEPARATOR = "=";
		private const string EMPTY_STRING = "";

		private readonly List<Parameter> parameters;

		public ParameterList()
		{
			parameters = new List<Parameter>();
		}

		internal ParameterList(IList<Parameter> parameters)
		{
			this.parameters = new List<Parameter>(parameters);
		}

		public ParameterList(IDictionary<string, string> map) : this()
		{
			foreach (KeyValuePair<string, string> entry in map)
				parameters.Add(new Parameter(entry.Key, entry.Value));
		}

		public virtual void add(string key, string value)
		{
			parameters.Add(new Parameter(key, value));
		}

		public virtual string appendTo(string url)
		{
			Debug.Assert(url != null, "Cannot append to null URL");

			string queryString = asFormUrlEncodedString();
			if (queryString.Equals(EMPTY_STRING))
				return url;
			else
			{
				url += url.IndexOf(QUERY_STRING_SEPARATOR_CHAR) != -1 ? PARAM_SEPARATOR : QUERY_STRING_SEPARATOR;
				url += queryString;
				return url;
			}
		}

		public virtual string asOauthBaseString()
		{
			return OAuthEncoder.encode(asFormUrlEncodedString());
		}

		public virtual string asFormUrlEncodedString()
		{
			if (parameters.Count == 0)
				return EMPTY_STRING;

			StringBuilder builder = new StringBuilder();
			foreach (Parameter p in parameters)
			{
				builder.Append('&');
				builder.Append(p.asUrlEncodedPair());
			}

			return builder.ToString().Substring(1);
		}

		public virtual void addAll(ParameterList other)
		{
			parameters.AddRange(other.parameters);
		}

		public virtual void addQueryString(string queryString)
		{
			if (queryString != null && queryString.Length > 0)
			{
				foreach (string param in queryString.Split(PARAM_SEPARATOR, true))
				{
					string[] pair = param.Split(PAIR_SEPARATOR, true);
					string key = OAuthEncoder.decode(pair[0]);
					string value = pair.Length > 1 ? OAuthEncoder.decode(pair[1]) : EMPTY_STRING;
					parameters.Add(new Parameter(key, value));
				}
			}
		}

		public virtual bool contains(Parameter param)
		{
			return parameters.Contains(param);
		}

		public virtual int size()
		{
			return parameters.Count;
		}

		public virtual ParameterList sort()
		{
			ParameterList sorted = new ParameterList(parameters);
			// TODO (Bret): This doesn't seem exactly right--the OAuth spec says that parameters should be sorted in byte order
			PlatformUtils.sortList(sorted.parameters);
			return sorted;
		}
	}

}

File: jsimple-oauth/src/main/java/jsimple/oauth/model/ParameterList.java
/*
 * Portions copyright (c) 2012-2014 Microsoft Mobile.  All Rights Reserved.
 *
 * This file is based on or incorporates material from the Scribe Java OAuth
 * library https://github.com/fernandezpablo85/scribe-java (collectively, "Third
 * Party Code"). Microsoft Mobile is not the original author of the Third Party
 * Code.
 *
 *
 * Permission is hereby granted, free of charge, to any person obtaining a copy
 * of this software and associated documentation files (the "Software"), to deal
 * in the Software without restriction, including without limitation the rights
 * to use, copy, modify, merge, publish, distribute, sublicense, and/or sell
 * copies of the Software, and to permit persons to whom the Software is
 * furnished to do so, subject to the following conditions:
 *
 * The above copyright notice and this permission notice shall be included in
 * all copies or substantial portions of the Software.
 *
 * THE SOFTWARE IS PROVIDED "AS IS", WITHOUT WARRANTY OF ANY KIND, EXPRESS OR
 * IMPLIED, INCLUDING BUT NOT LIMITED TO THE WARRANTIES OF MERCHANTABILITY,
 * FITNESS FOR A PARTICULAR PURPOSE AND NONINFRINGEMENT. IN NO EVENT SHALL THE
 * AUTHORS OR COPYRIGHT HOLDERS BE LIABLE FOR ANY CLAIM, DAMAGES OR OTHER
 * LIABILITY, WHETHER IN AN ACTION OF CONTRACT, TORT OR OTHERWISE, ARISING FROM,
 * OUT OF OR IN CONNECTION WITH THE SOFTWARE OR THE USE OR OTHER DEALINGS IN
 * THE SOFTWARE.
 */

package jsimple.oauth.model;

import jsimple.oauth.utils.OAuthEncoder;
import jsimple.util.PlatformUtils;
import org.jetbrains.annotations.Nullable;

import java.util.ArrayList;
import java.util.List;
import java.util.Map;

/**
 * @author: Pablo Fernandez
 */
public class ParameterList {
    private static final char QUERY_STRING_SEPARATOR_CHAR = '?';
    private static final String QUERY_STRING_SEPARATOR = "?";
    private static final String PARAM_SEPARATOR = "&";
    private static final String PAIR_SEPARATOR = "=";
    private static final String EMPTY_STRING = "";

    private final ArrayList<Parameter> parameters;

    public ParameterList() {
        parameters = new ArrayList<Parameter>();
    }

    ParameterList(List<Parameter> parameters) {
        this.parameters = new ArrayList<Parameter>(parameters);
    }

    public ParameterList(Map<String, String> map) {
        this();
        for (Map.Entry<String, String> entry : map.entrySet()) {
            parameters.add(new Parameter(entry.getKey(), entry.getValue()));
        }
    }

    public void add(String key, String value) {
        parameters.add(new Parameter(key, value));
    }

    public String appendTo(String url) {
        assert url != null : "Cannot append to null URL";

        String queryString = asFormUrlEncodedString();
        if (queryString.equals(EMPTY_STRING)) {
            return url;
        } else {
            url += url.indexOf(QUERY_STRING_SEPARATOR_CHAR) != -1 ? PARAM_SEPARATOR : QUERY_STRING_SEPARATOR;
            url += queryString;
            return url;
        }
    }

    public String asOauthBaseString() {
        return OAuthEncoder.encode(asFormUrlEncodedString());
    }

    public String asFormUrlEncodedString() {
        if (parameters.size() == 0)
            return EMPTY_STRING;

        StringBuilder builder = new StringBuilder();
        for (Parameter p : parameters) {
            builder.append('&');
            builder.append(p.asUrlEncodedPair());
        }

        return builder.toString().substring(1);
    }

    public void addAll(ParameterList other) {
        parameters.addAll(other.parameters);
    }

    public void addQueryString(@Nullable String queryString) {
        if (queryString != null && queryString.length() > 0) {
            for (String param : queryString.split(PARAM_SEPARATOR)) {
                String[] pair = param.split(PAIR_SEPARATOR);
                String key = OAuthEncoder.decode(pair[0]);
                String value = pair.length > 1 ? OAuthEncoder.decode(pair[1]) : EMPTY_STRING;
                parameters.add(new Parameter(key, value));
            }
        }
    }

    public boolean contains(Parameter param) {
        return parameters.contains(param);
    }

    public int size() {
        return parameters.size();
    }

    public ParameterList sort() {
        ParameterList sorted = new ParameterList(parameters);
        // TODO (Bret): This doesn't seem exactly right--the OAuth spec says that parameters should be sorted in byte order
        PlatformUtils.sortList(sorted.parameters);
        return sorted;
    }
}
